lib/api/src/externals/function.rs

```diff
@@ -478,16 +480,23 @@ impl Function {
         }
 
         // Call the trampoline.
-        if let Err(error) = unsafe {
-            wasmer_call_trampoline(
-                &self.store,
-                self.exported.vm_function.vmctx,
-                trampoline,
-                self.exported.vm_function.address,
-                values_vec.as_mut_ptr() as *mut u8,
-            )
-        } {
-            return Err(RuntimeError::from_trap(error));
+        unsafe {
+            if !unchecked {
+                wasmer_call_trampoline(
+                    &self.store,
+                    self.exported.vm_function.vmctx,
+                    trampoline,
+                    self.exported.vm_function.address,
+                    values_vec.as_mut_ptr() as *mut u8,
+                )?
+            } else {
+                wasmer_call_trampoline_unchecked(
```

Comment:
It may be possible to represent, in the typesystem, a token that allows calling a function in an unchecked way. The only way to get such a token would be from the `catch_traps` function. This would allow us to not have to mark these functions as unsafe because they'd only be callable in that context.

Here's what I'm thinking:

```Rust
/// A token indicating permission to call Wasm functions without an extra trap handler.
#[derive(Debug)] // this type must not be `Copy` or `Clone`
pub struct CallUncheckedToken(PhantomData<()>); // some zero-sized private internal value so no one can construct this type outside of the file that it's declared in.

fn call_unchecked<F>(f: F) where F: FnOnce(&CallUncheckedToken, ...) -> ... {
   let call_unchecked_token = CallUncheckedToken
   catch_traps(||  f.call_unchecked(&call_unchecked_token, ...))
}
```

And this ensures that all `call_unchecked`s happen in a context where a trap catcher exists. It does mean users have to pass that token around and I'm not 100% sure how trap handling works, does it matter if there are function calls on the stack between the `catch_traps` and the function call itself? If so, then this solution doesn't work, for example given a stack trace from newest to oldest:

```
some_wasm_function_that_traps
host_function2
host_function1
catch_traps
...
```

We'll unwind through those host functions to get to `catch_traps`. If that's not a problem, this solution works. If it is a problem, this solution does not work.